proxy/src/main/java/com/wavefront/agent/CachingGraphiteHostAnnotator.java

```diff
@@ -0,0 +1,69 @@
+package com.wavefront.agent;
+
+import com.google.common.collect.Lists;
+
+import com.github.benmanes.caffeine.cache.Caffeine;
+import com.github.benmanes.caffeine.cache.LoadingCache;
+import com.wavefront.metrics.ExpectedAgentMetric;
+import com.yammer.metrics.Metrics;
+import com.yammer.metrics.core.Gauge;
+
+import java.net.InetAddress;
+import java.net.InetSocketAddress;
+import java.util.List;
+import java.util.concurrent.TimeUnit;
+import java.util.stream.Collectors;
+
+import io.netty.channel.ChannelHandler;
+import io.netty.channel.ChannelHandlerContext;
+
+/**
+ * Given a raw Graphite/Wavefront line, look for any host tag, and add it if implicit.
+ *
+ * Differences from GraphiteHostAnnotator:
+ * - sharable
+ * - lazy load - does not proactively perform rDNS lookups unless needed
+ * - can be applied to HTTP payloads
+ *
+ * @author vasily@wavefront.com
+ */
+@ChannelHandler.Sharable
+public class CachingGraphiteHostAnnotator {
+  private final LoadingCache<InetAddress, String> rdnsCache;
+  private final boolean disableRdnsLookup;
+  private final List<String> sourceTags;
+
+  public CachingGraphiteHostAnnotator(final List<String> customSourceTags, boolean disableRdnsLookup) {
+    this.disableRdnsLookup = disableRdnsLookup;
+    this.sourceTags = Lists.newArrayListWithExpectedSize(customSourceTags.size() + 2);
```

Comment:
Could customSourceTags be null? If so, add null checks, ignore this comment otherwise.

Reply:
It's always initialized so it's never null in practice, but you have a very good point - null is a legitimate use case here. Added handling for null cases. 
And, of course, you're right - it should be +4.